lib/inspec/config.rb

```diff
@@ -0,0 +1,391 @@
+      credentials.each_with_object({}) do |(option, value), creds|
+        creds[option.to_sym] = value
+        creds
+      end
+    end
+
+    private
+
+    def _utc_merge_transport_options(credentials, transport_name)
+      # Ask Train for the names of the transport options
+      transport_options = Train.options(transport_name).keys.map(&:to_s)
+
+      # If there are any options with those (unprefixed) names, merge them in.
+      unprefixed_transport_options = final_options.select do |option_name, _value|
+        transport_options.include? option_name # e.g., 'host'
+      end
+      credentials.merge!(unprefixed_transport_options)
+
+      # If there are any prefixed options, merge them in, stripping the prefix.
+      transport_prefix = transport_name.downcase.tr('-', '_') + '_'
+      transport_options.each do |bare_option_name|
+        prefixed_option_name = transport_prefix + bare_option_name.to_s
+        if final_options.key?(prefixed_option_name)
+          credentials[bare_option_name.to_s] = final_options[prefixed_option_name]
+        end
+      end
+    end
+
+    # fetch any info that applies to all transports (like sudo information)
+    def _utc_generic_credentials
+      @final_options.select { |option, _value| GENERIC_CREDENTIALS.include?(option) }
+    end
+
+    def _utc_determine_backend(credentials)
+      return if credentials.key?(:backend)
+
+      # Default to local
+      unless @final_options.key?(:target)
+        credentials[:backend] = 'local'
+        return
+      end
+
+      # Look into target
+      %r{^(?<transport_name>[a-z_\-0-9]+)://.*$} =~ final_options[:target]
+      unless transport_name
+        raise ArgumentError, "Could not recognize a backend from the target #{final_options[:target]} - use a URI format with the backend name as the URI schema.  Example: 'ssh://somehost.com' or 'transport://credset' or 'transport://' if credentials are provided outside of InSpec."
+      end
+      credentials[:backend] = transport_name.to_s # these are indeed stored in Train as Strings.
+    end
+
+    #-----------------------------------------------------------------------#
+    #                         Reading Config Files
+    #-----------------------------------------------------------------------#
+
+    # Regardless of our situation, end up with a readable IO object
+    def resolve_cfg_io(cli_opts, cfg_io)
+      raise(ArgumentError, 'Inspec::Config must use an IO to read from') if cfg_io && !cfg_io.respond_to?(:read)
+      cfg_io ||= check_for_piped_config(cli_opts)
+      return cfg_io if cfg_io
+
+      path = determine_cfg_path(cli_opts)
+
+      cfg_io = File.open(path) if path
+      cfg_io || StringIO.new('{ "version": "1.1" }')
+    end
+
+    def check_for_piped_config(cli_opts)
+      cli_opt = cli_opts[:config] || cli_opts[:json_config]
+      Inspec.deprecate(:cli_option_json_config, '') if cli_opts.key?(:json_config)
+
+      return nil unless cli_opt
```

Comment:
Curious why both of these are needed. Wouldn't the second guard be sufficient because if it is nil it will evaluate to `false`?

Reply:
You're right, `nil == '-'` does evaluate to false, so we could rely on the second guard only. I didn't consider that, and we could make that change, in the spirit of brevity.

I think of guards like this: one by one, think of each reason you can get out of doing the work you were asked to do. So, here, I have two reasons: They didn't pass in a --config option at all (guard 1) or they did use a --config option but it was not '-'.  Two different reasons, so two different guards, for clarity.

I usually favor clarity over brevity, though sometimes the more concise version is clearer.

Reply:
Hmm...never heard that argument before. I really don't have an opinion one way or another given this new info :smile:. Not like this will affect performance.